core/src/apps/cardano/sign_tx.py

```diff
@@ -100,526 +111,529 @@
+        )
 
-def _cborize_inputs(inputs: list[CardanoTxInputType]) -> list[tuple[bytes, int]]:
-    return [(tx_input.prev_hash, tx_input.prev_index) for tx_input in inputs]
+    if output.address_parameters:
+        validate_address_parameters(output.address_parameters)
+    elif output.address is not None:
+        validate_output_address(output.address, msg.protocol_magic, msg.network_id)
+    else:
+        raise wire.ProcessError(
+            "Each output must have an address field or address_parameters!"
+        )
 
 
-def _cborize_outputs(
+async def _show_output(
+    ctx: wire.Context,
     keychain: seed.Keychain,
-    outputs: list[CardanoTxOutputType],
-    protocol_magic: int,
-    network_id: int,
-) -> list[CborizedTxOutput]:
-    return [
-        _cborize_output(keychain, output, protocol_magic, network_id)
-        for output in outputs
-    ]
+    msg: CardanoSignTxInit,
+    output: CardanoTxOutput,
+) -> int | None:
+    """Return the amount of the output if it should be shown to the user."""
 
-
-def _cborize_output(
-    keychain: seed.Keychain,
-    output: CardanoTxOutputType,
-    protocol_magic: int,
-    network_id: int,
-) -> CborizedTxOutput:
-    amount = output.amount
     if output.address_parameters:
-        address = derive_address_bytes(
-            keychain, output.address_parameters, protocol_magic, network_id
+        await _fail_or_warn_if_invalid_path(
+            ctx,
+            (SCHEMA_ADDRESS,),
+            output.address_parameters.address_n,
+            CHANGE_OUTPUT_PATH_NAME,
+        )
+
+        await _show_change_output_staking_warnings(
+            ctx, keychain, output.address_parameters, output.amount
         )
-    else:
-        assert output.address is not None  # _validate_outputs
-        address = get_address_bytes_unsafe(output.address)
 
-    if not output.token_bundle:
-        return (address, amount)
+        if _should_hide_output(output.address_parameters.address_n):
+            return None
+
+        address = derive_human_readable_address(
+            keychain, output.address_parameters, msg.protocol_magic, msg.network_id
+        )
     else:
-        return (address, (amount, _cborize_token_bundle(output.token_bundle)))
+        assert output.address is not None  # _validate_output
+        address = output.address
 
+    if output.asset_groups_count > 0:
+        await show_warning_tx_output_contains_tokens(ctx)
 
-def _cborize_token_bundle(
-    token_bundle: list[CardanoAssetGroupType],
-) -> CborizedTokenBundle:
-    result: CborizedTokenBundle = {}
+    if msg.signing_mode == CardanoTxSigningMode.ORDINARY_TRANSACTION:
+        await confirm_sending(ctx, output.amount, address)
+        return output.amount
 
-    for token_group in token_bundle:
-        cborized_policy_id = bytes(token_group.policy_id)
-        cborized_token_group = result[cborized_policy_id] = {}
+    return None
```

Comment:
It's not clear to me what cases should end with this. Maybe some explanatory comments along the way would help, and an assert before this that we really want to reach it?

Reply:
The return value has been removed since we no longer need `total_amount_to_show`.